cloud/softlayer/sl_metal.py

```diff
@@ -0,0 +1,906 @@
+        items = self.get_items('id', package_id, dcenter)
+
+        item_cats = dict( 
+            (c.get('itemCategory').get('categoryCode'), [ 
+                i.get('keyName') for i in items.itervalues() \
+                if i.get('itemCategory').get('id') == c.get('itemCategory').get('id') 
+            ]) for c in cats
+        )
+
+        # Massage the disk options since the SoftLayer API is inconsistent and broken
+        disk_opts = item_cats.get('disk0')
+        for k in item_cats:
+            if re.match('disk[0-9]+', k):
+                item_cats[k] = disk_opts
+
+        return item_cats
+
+
+class Order(object):
+    """
+    Order - SoftLayer product order class
+    """
+    def __init__(self, client, module):
+        self.client = client
+        self.module = module
+
+        self.stdparams = [
+            'name',
+            'hostname',
+            'domain',
+            'hourly',
+            'pkgid',
+            'datacenter',
+            'primaryVlan',
+            'backendVlan',
+            'imageTemplateId',
+            'sshKeys',
+            'storageGroups',
+            'state',
+        ]
+
+        # Get the list it itemkeys to get prices for
+        self.item_values = [ module.params.get(x) for x in module.params if x not in self.stdparams ]
+
+        for k in self.stdparams:
+            setattr(self, k, module.params.get(k))
+
+    def getLocationId(self, location):
+        """Return the id of the datacenter specified by its short name."""
+        filt = {'name': {'operation': location}}
+        locations = self.client['Location'].getDataCenters(mask='id, name', filter=filt)
+        if len(locations) != 1:
+            return None     
+        return locations[0].get('id')
+
+    def getProductOrder(self):
+        """Return the product order in a structure consumable by the SL API"""
+        sld = SL_data(self.client)
+       
+        # For each item key get the price id, unless it is not specified
+        prices = [ 
+            {'id': sld.get_item_priceid(x,self.pkgid, self.datacenter), 'name':x} for x in self.item_values if x is not None 
+        ]
+        
+        productOrder = {
+            "quantity": 1,
+            "hardware": [{
+                            "hostname": self.hostname,
+                            "domain": self.domain,
+                            "primaryBackendNetworkComponent": self.backendVlan,
+                            "primaryNetworkComponent": self.primaryVlan
+                        }],
+            "imageTemplateId": self.imageTemplateId,
+            "location": self.getLocationId(self.datacenter),
+            "useHourlyPricing": self.hourly,
+            "packageId": self.pkgid,
+            "prices": prices,
+            "sshKeys": [self.sshKeys],
+            "storageGroups": [self.storageGroups]
+        }
+        return productOrder
+
+
+def main():
+    client = SoftLayer.create_client_from_env()
```

Comment:
Since you already defined `HAS_SL`, it seems like you might want to add a `fail_json` before this statement with a custom error message if SoftLayer can't be imported. I say this because it looks like other modules follow this pattern, and because the variable isn't otherwise used.